README.md
# supermockapi
[![Build Status](https://travis-ci.org/gegana/supermockapi.svg?branch=master)](https://travis-ci.org/gegana/supermockapi)
[![This image on DockerHub](https://img.shields.io/docker/pulls/dsumiskum/supermockapi.svg)](https://hub.docker.com/r/dsumiskum/supermockapi/)
[![npm version](https://badge.fury.io/js/supermockapi.svg)](https://badge.fury.io/js/supermockapi)

A lightweight, user-friendly, and powerful tool for mocking API endpoints. Useful for testing API client integrations, or performance testing your application without sending traffic to your API dependencies.
Has an external dependency to a mongo database.

# Running in Docker
Runs in port 3000 by default.
```shell
docker run -d --name supermockapi -p 80:3000 dsumiskum/supermockapi
```
Configure a different port.
```shell
docker run -d --name supermockapi -p 80:5000 -e "PORT:5000" dsumiskum/supermockapi
```
Specify initial mocked api routes specification through volume mapping a folder with a `routes.json` or the file itself to the container workdir (app is the workdir).
The application will look for a routes.json file in the workdir when booting up.
```shell
docker run -d --name supermockapi -p 3000:3000 -v "${pwd}/folder/routes.json:/app/routes.json" dsumiskum/supermockapi
```
Route specifications format (routes.json):
```json
[
  {
    "path": "/product/:sku", 
    "method": "GET",
    "status": 200, 
    "body": {
      "message": "Server received your request for sku {sku}" 
    },
    "responseDelay": 5000 
  },
  {
    "path": "/product/:sku", 
    "method": "POST",
    "status": 500
  }
]
```

# Features
supermockapi is meant to be lightweight, easy to deploy, and works for your Continous Integration needs. A big problem with performance testing 
services is isolating the external dependencies to get more predictable metrics, which would be useful if you are benchmarking your metrics
overtime.

With supermockapi, you can deploy it with your application as a stack, with specified routes.json file. And/or you can use the super friendly UI to
manually add mock api endpoints. The UI comes with a console log stream, so that you can monitor incoming requests all in one place.

## Create mocked API endpoints using Express JS pattern matching
You can do this via the UI or by configuring a routes.json file. Json file example where you take a parameter sku in the base path:
```json
{
  "path": "/product/:sku"
}
```

## Specify default response code
You can do this via the UI or by configuring a routes.json file. Json file example for returning a 200:
```json
{
  "path": "/product/:sku",
  "status": 200
}
```

## Specify default response body in json
You can do this via the UI or by configuring a routes.json file. Json file example for returning a simple message:
```json
{
  "path": "/product/:sku",
  "status": 200,
  "body": {
    "message": "Hello world!"
  }
}
```
You can include values from the tokens bag in your response body. The syntax for this is `{token}`.
```json
{
  "path": "/product/:sku",
  "status": 200,
  "body": {
    "sku": "{sku}"
  }
}
```
What is included in tokens:
1. Any parameters in the base URL
2. Any parameters in the querystring
3. The request body `{body}`
4. The number of calls made to the base URL `{calls}`
5. A random number between 1 and 100 `{random}`
6. The current timestamp `{timestamp}`

## Simulate response delay
You can do this via the UI or by configuring a routes.json file. Json file example for delaying response by 5 seconds:
```json
{
  "responseDelay": 5000
}
```

## Specify conditional behaviors
You can do this via the UI or by configuring a routes.json file. Json file example of a weighted response scenario using conditional behaviors:
```json
{
    "path": "/random",
    "method": "GET",
    "status": 200,
    "conditionalBehaviors": [{
        "condition": "{random} <= 30",
        "status": 500,
        "body": {
          "message": "[Condition Triggered] Random = {random}"
        }
      },
      {
        "condition": "{random} <= 60",
        "status": 404,
        "body": {
          "message": "[Condition Triggered] Random = {random}"
        }
      },
      {
        "condition": "{random} <= 90",
        "status": 200,
        "body": {
          "message": "[Condition Triggered] Random = {random}"
        }
      },
      {
        "condition": "{random} <= 100",
        "status": 206,
        "body": {
          "message": "[Condition Triggered] Random = {random}"
        }
      }
    ],
    "body": {
      "message": "Number of calls = {calls}"
    }
  }
```

# How it works
Built using express js, the mock api endpoints have 3 priority levels:
1. Routes hard coded in the application
2. Routes you inject via routes.json
3. Routes added dynamically via the API (/mockapi/route) that the UI uses

The routes.json configurations are in memory, while the routes added dynamically via the UI are stored in mongodb.
When you create routes.json file, make sure to include the required fields: path, method, and status.

# Contributing
If you want to contribute, the application is fairly standard node js app albeit it is using 7.0 features for async/await operators. So you will need:
1. Node 7.10.0
2. IDE like Visual Studio Code
3. Docker engine


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.3.0"></a>
# [1.3.0](https://github.com/gegana/supermockapi/compare/v1.2.3...v1.3.0) (2017-12-14)


### Features

* replace mongodb with lokijs ([589393f](https://github.com/gegana/supermockapi/commit/589393f))



<a name="1.2.3"></a>
## [1.2.3](https://github.com/gegana/supermockapi/compare/v1.2.2...v1.2.3) (2017-07-11)



<a name="1.2.2"></a>
## [1.2.2](https://github.com/gegana/supermockapi/compare/v1.2.1...v1.2.2) (2017-05-27)


### Bug Fixes

* use base Docker image node:7.10-alpine to reduce image size. ([1a65c10](https://github.com/gegana/supermockapi/commit/1a65c10))



<a name="1.2.1"></a>
## [1.2.1](https://github.com/gegana/supermockapi/compare/v1.2.0...v1.2.1) (2017-05-25)


### Bug Fixes

* req.body keys was not flattened in tokens. ([44bd842](https://github.com/gegana/supermockapi/commit/44bd842))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/gegana/supermockapi/compare/v1.1.5...v1.2.0) (2017-05-25)


### Features

* Conditional behaviors support. Tokens now include calls, random, timestamp, and req body. Updated UI to support conditional behaviors. ([dc0ea9e](https://github.com/gegana/supermockapi/commit/dc0ea9e))



<a name="1.1.5"></a>
## [1.1.5](https://github.com/gegana/supermockapi/compare/v1.1.4...v1.1.5) (2017-05-23)


### Bug Fixes

* Big rename from mockapi to supermockapi. ([dd825f5](https://github.com/gegana/supermockapi/commit/dd825f5))



<a name="1.1.4"></a>
## [1.1.4](https://github.com/gegana/mockapi/compare/v1.1.3...v1.1.4) (2017-05-23)


### Bug Fixes

* fixing npm publish error in travis ci by changing npm package name to supermockapi. ([9666c6a](https://github.com/gegana/mockapi/commit/9666c6a))



<a name="1.1.3"></a>
## [1.1.3](https://github.com/gegana/mockapi/compare/v1.1.2...v1.1.3) (2017-05-23)


### Bug Fixes

* fixing npm publish error in travis ci by changing npm package name to mock-api. ([41351e1](https://github.com/gegana/mockapi/commit/41351e1))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/gegana/mockapi/compare/v1.1.1...v1.1.2) (2017-05-23)


### Bug Fixes

* fixing npm publish error in travis ci. ([c9fa879](https://github.com/gegana/mockapi/commit/c9fa879))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/gegana/mockapi/compare/v1.1.0...v1.1.1) (2017-05-23)


### Bug Fixes

* make package.json public. ([2cada07](https://github.com/gegana/mockapi/commit/2cada07))



<a name="1.1.0"></a>
# 1.1.0 (2017-05-23)


### Features

* introduces stateless mode. npm publish for travis. can now specify routes.json file by absolute path. ([3918087](https://github.com/gegana/mockapi/commit/3918087))


src/startup.js
"use strict"

const Timestamp = require('./lib/timestamp'),
  RouteFactory = require('./routeFactory'),
  RouteHandler = require('./services/routeHandler'),
  Hyperwatch = require('hyperwatch'),
  Session = require('express-session'),
  RouteRepository = require('./data/routeRepository')

/**
 * Bootstrap express js application
 */
module.exports = class Startup {

  /**
   * Default HTTP routes
   */
  get routes() {
    return {
      'index': require('./routes/index'),
      'api': require('./routes/api'),
      'gateway': require('./routes/gateway')
    }
  }

  /**
   * Default empty constructor
   */
  constructor() {
    this.__routeFactory = new RouteFactory()
    this.__config = require('../config.json')
  }

  /**
   * Configure service dependencies
   * @param {object} services 
   */
  useServices(services) {
    for (let key in services) {
      this.__routeFactory.useService(key, services[key])
    }
    return this
  }

  /**
   * Boot an express js application
   * @param {object} routes 
   * @param {string} port 
   * @param {bool} stateless - Stateless mode means no UI, API, and persistence storage
   */
  async boot(routes, port) {
    console.log('[' + Timestamp() + '] Booting application')
    await this.initApp(routes)
    this.start(port)
  }

  /**
   * Construct an express js app with minimal configuration
   * @param {object} routes
   * @param {bool} stateless - Stateless mode means no UI, API, and persistence storage 
   */
  async initApp(routes) {
    console.log('[' + Timestamp() + '] Initializing express js application')
    this.app = require('express')()
    this.setView('twig')
      .setParsers()
      .setLogger()
      .useSession()
      .useRouteRepo()
      .setRoutes(routes)
    return this
  }

  /**
   * Start express js app running on specified port
   * @param {string} port 
   */
  start(port) {
    var port = port || '3000'
    console.log('[' + Timestamp() + '] Starting server on port ' + port)
    this.app.set('port', port)
    this.server = require('http').createServer(this.app)
    this.server.listen(port)
    this.server.on('error', this.__onError)
    this.server.on('listening', this.__onListening)
    Hyperwatch(this.server)
  }

  /**
   * Event listener for server errors
   * @param {object} err 
   */
  __onError(err) {
    if (error.syscall !== 'listen') {
      throw error
    }

    var bind = typeof port === 'string' ?
      'Pipe ' + port :
      'Port ' + port

    // handle specific listen errors with friendly messages
    switch (error.code) {
      case 'EACCES':
        console.error(bind + ' requires elevated privileges')
        process.exit(1)
        break
      case 'EADDRINUSE':
        console.error(bind + ' is already in use')
        process.exit(1)
        break
      default:
        throw error
    }
  }

  /**
   * Event listener for when server starts listening to HTTP port
   */
  __onListening() {
    const debug = require('debug')('mockapi:server')
    var addr = this.address()
    var bind = typeof addr === 'string' ?
      'pipe ' + addr :
      'port ' + addr.port
    debug('Listening on ' + bind)
  }

  /**
   * Setup express js app to use a particular view engine
   * @param {string} viewEngine 
   */
  setView(viewEngine) {
    // view engine setup
    console.log('[' + Timestamp() + '] Setting express js view engine')
    this.app.set('views', require('path').join(__dirname, 'views'))
    this.app.set('view engine', viewEngine)
    return this
  }

  /**
   * Setup express js app to use morgan logger
   */
  setLogger() {
    console.log('[' + Timestamp() + '] Setting express js logger')
    const morgan = require('morgan')
    this.app.use(morgan('dev'))
    return this
  }

  /**
   * Setup express js cookie and json parser
   */
  setParsers() {
    console.log('[' + Timestamp() + '] Setting express js HTTP header and content parsers')
    const cookieParser = require('cookie-parser'),
      bodyParser = require('body-parser')

    this.app.use(bodyParser.json())
    this.app.use(bodyParser.urlencoded({
      extended: false
    }))
    this.app.use(cookieParser())

    return this
  }

  /**
   * Setup express js to use session - can be accessed from req.session
   */
  useSession() {
    this.app.use(Session({
      secret: 'moistened weasels',
      resave: false,
      saveUninitialized: true
    }))
    return this
  }

  /**
   * Setup route repository
   */
  useRouteRepo() {
    const routeRepo = new RouteRepository()
    this.__routeFactory.tryUseService('routeRepository', routeRepo)
    return this
  }

  /**
   * Setup express js app HTTP routes
   * @param {object} routes 
   * @param {bool} stateless - Stateless mode means no UI, API, and persistence storage 
   */
  setRoutes(routes) {
    console.log('[' + Timestamp() + '] Setting express js routes')
    const globalErrorHandler = require('./middlewares/globalErrorHandler'),
      pageNotFoundHandler = require('./middlewares/pageNotFoundHandler')

    // Handle static file requests to public folder
    this.app.use(
      require('express')
      .static(require('path')
        .join(__dirname, 'public')
      )
    )

    // Setup app to use defined routes
    for (let key in this.routes) {
      let route = this.routes[key]
      if (Array.isArray(route)) {
        for (let key in route) {
          var r = route[key]
          this.__routeFactory.register(r.method, r.path, r.handler)
        }
      } else {
        this.__routeFactory.register(route.method, route.path, route.handler)
      }
    }

    // Setup injected routes
    var routes = routes || {}
    for (let key in routes) {
      let route = routes[key]
      this.__routeFactory.register(route.method, route.path, function (req, res, next) {
        const h = new RouteHandler(route)
        h.processRequest(req, res, next).then().catch((err) => {
          next(err)
        })
      })
    }

    this.app.use(this.__routeFactory.router)

    // Default to 404 when route not found
    this.app.use(pageNotFoundHandler)

    // Handle errors globally
    this.app.use(globalErrorHandler)

    return this
  }
}